torch/fx/experimental/proxy_tensor.py

```diff
@@ -158,6 +119,9 @@ def proxy_call(proxy_mode, func_overload, args, kwargs=None):
         if r is not NotImplemented:
             return r
 
+    # DODGY af
+    g = DisableTracingGuard(proxy_mode.sym_mode)
```

Comment:
I'm a little confused why we need this--if the SymIntMode always run before this, shouldn't the current SymIntMode be None? If it's run after, I'm not sure why we're not tracing

Reply:
Ah--missed an update

Reply:
So, this is out of date with the most recent push, but the underlying point is still valid, let me try to address it.

The crux of the issue is that we have separate modes for SymInt and for Tensors. That means there are two distinct pieces of state we have to deal with. If we enter the TorchDispatchMode, the TorchDispatchMode will be suspended... but the SymDispatchMode will not be. If we want SymDispatchMode to also be disabled, we have to do it explicitly.

We could have sidestepped this by only having one mode. I considered this. What seemed troublesome, however, was it meant that you could end up calling TorchDispatchMode with operators that are not `torch.ops` operators (since the SymInt operators are things from the `operators` module.) Maybe this is not so bad. Definitely open to other opinions about this.

Reply:
> but the underlying point is still valid, let me try to address it

Thanks for calling that out

> If we want SymDispatchMode to also be disabled, we have to do it explicitly.

To check my understanding, we want this because we don't want to trace both the torch.ops and operators for the same function call?

> What seemed troublesome, however, was it meant that you could end up calling TorchDispatchMode with operators that are not `torch.ops` operators

One more question: the SymInt `operators` ops won't ever get caught by `__torch_dispatch__`, right? The issues is just that ProxyTensorMode would also end up being extended to catch `operators` ops

If that's the case, IMO it seems nice to have it taken care of as one mode that doesn't know about other modes. What if we left ProxyTensorMode as it is and inherited from it to add this new functionality (ProxyTensorAndSymIntMode??)? We could override enter/exit with super + the new behavior. Still imperfect but at least separates the behavior

Reply:
> To check my understanding, we want this because we don't want to trace both the torch.ops and operators for the same function call?

The mental model @Chillee taught me is that the proxy tensor FX trace should only characterize compute from the user code. So if a user writes `view(2 * x.shape[0])`, we ought to capture the multiplication on the symbolic shape, but if the user writes `convolution(...)`, that should show up as only a convolution. However, the way SymInts are implemented, internally we still do all of the SymInt computation. So we want to mask it out from the graph.

> One more question: the SymInt operators ops won't ever get caught by __torch_dispatch__, right? The issues is just that ProxyTensorMode would also end up being extended to catch operators ops

That's correct.

> If that's the case, IMO it seems nice to have it taken care of as one mode that doesn't know about other modes. What if we left ProxyTensorMode as it is and inherited from it to add this new functionality (ProxyTensorAndSymIntMode??)? We could override enter/exit with super + the new behavior. Still imperfect but at least separates the behavior

The main problem is that the mode class doesn't actually take care of the disabling: the internal C++ code that calls into the dispatch mode does the disabling. So there's no logical place to add the extra functionality...

Reply:
> The mental model @Chillee taught me is that the proxy tensor FX trace should only characterize compute from the user code. So if a user writes view(2 * x.shape[0]), we ought to capture the multiplication on the symbolic shape, but if the user writes convolution(...), that should show up as only a convolution. However, the way SymInts are implemented, internally we still do all of the SymInt computation. So we want to mask it out from the graph.

That explanation's super helpful thanks! Okay I'm sold on the current approach